### ISSUE TITLE
Batching changes made via Object.assign?

### ISSUE THREAD

**Issue by dyoder** [gullerya/object-observer#51]
Thank you for this excellent library.

My scenario is that I'm using `Object.assign`, which is (correctly) invoking the callback for each implicit assignment. So if I have an object with two properties, the callback is invoked twice.

However, from the application standpoint, this is one change. Is there some way to detect this? Or group the changes? The argument to the callback is an array, but always appears to have just one value.

I'm guessing (I have not studied the code) that this is just a set trap which then invokes the callback. I don't have any idea how you might group changes together. But I thought perhaps this might be something to which you've given some thought. Is there a way to do this? If not, is there a good workaround?

**Comment by gullerya** [comment]
Hi!
First - thanks for a feedback and really sorry for me missing this issue for quite a long time :(

Sounds right - I'll run some local tests and see what happens and how change might be batched in this case, if any.

Will report here ASAP.

**Comment by gullerya** [comment]
I've played with it a bit and the following seems like my current standpoint on this:
- indeed, `Object.assign` performs it's action in a sequence, one op. at a time, natively; outcome of this is, that a traps on the level of operated object/proxy can NOT be aware of the fact that they are invoked as part of a higher level batch operation/context
- overriding the `Object.assign` self is NOT a path, that I'd like to explore at this point - looks wrong to me from a several aspects
- possible solution to make it batch, is to actually postpone callbacking to the end of the current event loop, collect all events to that point and then perform a single dispatch to all events collected to the point

I'm thinking about the last point as the most correct path to take, this is in fact the way the native `MutationObserver` works.

Yet, an  inevitable outcome of this solution is that within that single callback may happen to be found not only the events of the single operation (eg those 2 mutations made be `Object.assign` from OP), but also another changes.

I'd like to think and experiment more about this, since I'm still not sure about this approach.

I don't think that there is any workaround as of now, I'd definitely NOT advise to manage any outer state/context of mutation operation - too fragile, IMHO.

In general, those events, be them delivered in a single callback each or grouped by, should be treated as a single continuous stream of events, so your application should not make any assumption on the delivery fashion, except that: any change will be notified, changes are notified in the order of execution.

Any insights are welcome.

**Comment by gullerya** [comment]
Example of a POSSIBLE async/batching behaviour (this is NOT HOW IT WORKS NOW, DON'T copy/paste :) ):
```
// SETUP
let counter = 0;
const oo = Observable.from({});
oo.observe(changes =>
    changes.forEach(() =>
        counter++;
    )
);

// MUTATIONS
Object.assign(oo, { a: 1, b: 2, c: 3 });
counter === 0; // true, no callbacks ran yet
oo.c = 4;
counter === 0; // true, no callbacks ran yet

// END OF EVENT LOOP
counter === 4; // true, only now the callbacks have run
```

**Comment by dyoder** [comment]
This sounds right. If they can be batched at the end of the event loop, as is done with MutationObserver, that would do the trick. That probably should have been the framing of the original ticket. As you say, the interface is what matters. With `Object.assign` we are in fact potentially making multiple changes. It doesn't matter if that was done by multiple assignments or one call to `Object.assign`. The real issue is that they are not batched. Thanks for your work on this!

**Comment by gullerya** [comment]
Thanks for the response, I'll start playing with the code around to see how it goes. 👍

**Comment by gullerya** [comment]
Related to #45 : the async approach mentioned there is the enabler for the functionality requested here.

**Comment by gullerya** [comment]
Since an async changes delivery is a real topic of this issue and an enabler for it, please refer to [this issue's details](https://github.com/gullerya/object-observer/issues/45#issuecomment-683655809) for an API and implementation details.

**Comment by gullerya** [comment]
implemented in 3.2.0
